// @fast-check/vitest/lib/cjs/internals/TestAlongGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testAPIRefined = void 0;
const suite_1 = require("vitest/suite");
const fast_check_1 = require("fast-check");
function isSig1OrSig2(args) {
    return typeof args[1] === 'function' || (args[1] === undefined && args[2] === undefined);
}
function taskCollectorBuilder(...args) {
    const [name, fn, options] = isSig1OrSig2(args) ? args : [args[0], args[2], args[1]];
    const taskName = typeof name === 'function' ? name.name : name;
    (0, suite_1.getCurrentSuite)().task(taskName, {
        ...this,
        ...(typeof options === 'number' ? { timeout: options } : options),
        handler: fn !== undefined
            ? async (context) => {
                let calledOnce = false;
                const config = (0, fast_check_1.readConfigureGlobal)();
                try {
                    const parameters = {
                        numRuns: config.numRuns ?? 1,
                        endOnFailure: config.endOnFailure ?? true,
                        includeErrorInReport: false,
                        errorWithCause: true,
                    };
                    await (0, fast_check_1.assert)((0, fast_check_1.asyncProperty)((0, fast_check_1.gen)(), (g) => {
                        const refinedG = Object.assign((arb, ...args) => {
                            calledOnce = true;
                            return g(arb, ...args);
                        }, { values: () => g.values() });
                        return fn({ ...context, g: refinedG });
                    }), parameters);
                }
                catch (error) {
                    if (calledOnce) {
                        throw error;
                    }
                    throw error.cause;
                }
            }
            : undefined,
    });
}
exports.testAPIRefined = (0, suite_1.createTaskCollector)(taskCollectorBuilder);
